vpr/src/pack/post_routing_pb_pin_fixup.cpp

```diff
@@ -0,0 +1,269 @@
+            rr_node_type = IPIN;
+        }
+
+        std::vector<e_side> pin_sides = find_logic_tile_pin_side(physical_tile, physical_pin);
+        /* As some grid has height/width offset, we may not have the pin on any side */
+        if (0 == pin_sides.size()) {
+            continue;
+        }
+
+        /* Special for I/O grids: VPR creates the grid with duplicated pins on every side 
+         * but the expected side (only used side) will be opposite side of the border side!
+         */
+        if (NUM_SIDES != border_side) {
+            pin_sides.clear();
+            switch (border_side) {
+                case TOP:
+                    pin_sides.push_back(BOTTOM);
+                    break;
+                case RIGHT:
+                    pin_sides.push_back(LEFT);
+                    break;
+                case BOTTOM:
+                    pin_sides.push_back(TOP);
+                    break;
+                case LEFT:
+                    pin_sides.push_back(RIGHT);
+                    break;
+                default:
+                    VPR_FATAL_ERROR(VPR_ERROR_ROUTE, "Invalid side for logic blocks at FPGA border!\n");
+            }
+        }
+
+        for (const e_side& pin_side : pin_sides) {
+            /* Find the net mapped to this pin in routing results */
+            const int& rr_node = get_rr_node_index(device_ctx.rr_node_indices,
+                                                   grid_coord.x(), grid_coord.y(),
+                                                   rr_node_type, physical_pin, pin_side);
+
+            /* Bypass invalid nodes, after that we must have a valid rr_node id */
+            if (-1 == rr_node) {
+                continue;
+            }
+            VTR_ASSERT((size_t)rr_node < device_ctx.rr_nodes.size());
+
+            /* Get the cluster net id which has been mapped to this net */
+            ClusterNetId routing_net_id = routing_ctx.rr_node_nets[RRNodeId(rr_node)];
+
+            /* Find the net mapped to this pin in clustering results*/
+            ClusterNetId cluster_net_id = clustering_ctx.clb_nlist.block_net(blk_id, j);
+
+            /* Ignore those net have never been routed */
+            if ((ClusterNetId::INVALID() != cluster_net_id)
+                && (true == clustering_ctx.clb_nlist.net_is_ignored(cluster_net_id))) {
+                continue;
+            }
+
+            /* Ignore used in local cluster only, reserved one CLB pin */
+            if ((ClusterNetId::INVALID() != cluster_net_id)
+                && (0 == clustering_ctx.clb_nlist.net_sinks(cluster_net_id).size())) {
+                continue;
+            }
+
+            /* If matched, we finish here */
+            if (routing_net_id == cluster_net_id) {
+                continue;
+            }
+
+            /* Update the clustering context with net modification */
+            clustering_ctx.post_routing_clb_pin_nets[blk_id][j] = routing_net_id;
```

Comment:
Indeed. I tried to modify the cluster nets but have no luck yet. I would like to hear some advice on which API to use. I have tried `set_pin_net()` and `associate_pin_with_net()` but they all give me segmentation fault. I looked into the codes, it seems to stash the nets and fail when rebuilding them. If you can give me some hints, I will try this further.

Reply:
> I have tried `set_pin_net()` and `associate_pin_with_net()` but they all give me segmentation fault. I looked into the codes, it seems to stash the nets and fail when rebuilding them. If you can give me some hints, I will try this further.

Hmm that is an issue. I'm trying to recall whether we actually ever fully implemented those APIs I know they are/were implemented for the AtomNetlist, but it is possible we didn't flesh them out in the ClusteredNetlist, since we didn't have any code which modified it at the time.

Reply:
No worries. I will keep these codes and try to finish regression tests. Then, we see how to streamline the data structure or result storage.